### PR TITLE
Added esConsumes to VirtualJetProducer classes

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- changed reco::writeSpecific jet helper functions to no longer take an EventSetup
- have VirtualJetProducer base class do esConsumes and added protected function to get the ES data products.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Please test

**Comment by Dr15Jones** [comment]
please test

**Review by jpata — COMMENTED** [review]
Thanks, looks good from the reco side (no changes). The usual minor comment...

File: RecoJets/JetProducers/src/JetSpecific.cc [hunk 1]
```diff
@@ -35,19 +33,19 @@ void reco::writeSpecific(reco::CaloJet& jet,
...
-                         edm::EventSetup const& c) {
+                         CaloGeometry const& geometry,
+                         HcalTopology const& topology) {
   // Get geometry
-  edm::ESHandle<CaloGeometry> geometry;
-  c.get<CaloGeometryRecord>().get(geometry);
-  const CaloSubdetectorGeometry* towerGeometry =
-      geometry->getSubdetectorGeometry(DetId::Calo, CaloTowerDetId::SubdetId);
+  //edm::ESHandle<CaloGeometry> geometry;
+  //c.get<CaloGeometryRecord>().get(geometry);
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply RecoJets/JetProducers/src/JetSpecific.cc:40]: Should go away

File: RecoJets/JetProducers/src/JetSpecific.cc [hunk 2]
```diff
@@ -35,19 +33,19 @@ void reco::writeSpecific(reco::CaloJet& jet,
...
+  const CaloSubdetectorGeometry* towerGeometry = geometry.getSubdetectorGeometry(DetId::Calo, CaloTowerDetId::SubdetId);
 
-  edm::ESHandle<HcalTopology> topology;
-  c.get<HcalRecNumberingRecord>().get(topology);
+  //edm::ESHandle<HcalTopology> topology;
+  //c.get<HcalRecNumberingRecord>().get(topology);
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply RecoJets/JetProducers/src/JetSpecific.cc:44]: Should go away

**Comment by Dr15Jones** [comment]
please test

**Comment by jpata** [comment]
+reconstruction
- technical, no reco changes expected/observed

**Comment by qliphy** [comment]
@Dr15Jones   Can you take care of CMS StaticAnalyzer warnings?:
There are 4 EventSetupRecord::get warnings. See https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bbaa8e/18712/llvm-analysis/esrget-sa.txt for details.

**Comment by Dr15Jones** [comment]
please test

**Comment by jpata** [comment]
+reconstruction
- as in https://github.com/cms-sw/cmssw/pull/35276#issuecomment-923038333

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
merge

**Comment by santocch** [comment]
+1

**Pull request merged**